### ISSUE TITLE
您好，想请教一下您一些问题

### ISSUE THREAD

**Issue by zzz-0930** [Lx-Bao/QSFNet#4]
十分感谢您的开源工作，拜读了您的文章，感慨其中的工作量之大！！
想请教下您训练的model，modelQA，modelT分别是代表什么，在我的理解中modelTY因该是最后得到的输出模型，modelQA是由质量感知区域选择子网训练获得的预测图，model则是区域引导选择性融合子网结果得到的预测图，不知道对不对。
另外我训练model和modelQA时batchsize可以设置为16，但训练modelT时batchsize设置为4就爆显存错误了，我的显卡是3090，不知道有什么方法去轻量化训练模型
最后想请教您的是我该怎么去测试我所训练的model以及meodlQA，以及如何输出它们的结果。
希望得到您的回复，十分感激！！

**Comment by zzz-0930** [comment]
您好，可以回复一下吗

**Comment by Lx-Bao** [comment]
您好，非常感谢您对我们论文的关注。
首先，model是Initial Feature Extraction Subnet，得到的是每个模态的初始特征及初步预测图以生成代表高低质量区域的伪GT；modelQA为Quality-aware Region Selection Subnet，目的是学习到高低质量区域；modelT则是Initial Feature Extraction Subnet和Region-guided Selective Fusion Subnet的结合，其融合多模态特征并生成最终预测图。
第二，您可以尝试使用NVIDIA Apex库来加速训练及降低显存占用，或者如果您有多卡的话可以使用多卡分布式训练。才外，在第三阶段训练时请记得冻结Quality-aware Region Selection Subnet的参数。
第三，您可以通过以下代码来得到model的结果，请点击[此处](https://pan.baidu.com/s/1QEAtWSr55ZrvpdUH5QGsXw?pwd=jz4c)下载，测试QAmodel的代码也是相似的。我们的评价指标的代码采用的是[SINet](https://github.com/DengPingFan/SINet)中的[EvaluationTool](https://github.com/DengPingFan/SINet/tree/master/EvaluationTool)，您可以使用matlab来运行这套评价代码。

**Comment by zzz-0930** [comment]
十分感谢您的答复，我按照您的建议二修改了Ttrain.py 在里面添加了torch.cuda.amp 以及禁用了qnet的参数：训练时间也从原来的300h变为了150h左右，但我感觉这个时间还是不太对，不知道您当时训练的时间是多少，我的model和modelQA都是用train-all进行训练的，大都在10h，20h。
我的环境配置是 RTX 3090  cuda 12.7  torch 2.3.1+cu121
以下是我修改后的Ttrain.py代码：

import os
from tqdm import tqdm
import torch
import numpy as np
from torch.utils.data import DataLoader
from lib.dataset import Data
from lib.data_prefetcher import DataPrefetcher
from torch.nn import functional as F
import pytorch_iou
from torch import nn
from QAnet import QAnet
from Tnet import Tnet
from torch.cuda.amp import autocast, GradScaler  # Import AMP

'''
第三阶段
'''

os.environ["CUDA_VISIBLE_DEVICES"] = '0'

torch.backends.cudnn.benchmark = True 
def bce_loss(pred, target):  # 第一个损失函数 bce 二分类交叉熵损失 主要是去预测前景和背景
    bce = F.binary_cross_entropy_with_logits(pred, target, reduction='mean')
    return bce

def structure_loss(pred, mask):
    weit = 1 + 5 * torch.abs(F.avg_pool2d(mask, kernel_size=31, stride=1, padding=15) - mask)
    wbce = F.binary_cross_entropy_with_logits(pred, mask, reduction='none')
    wbce = (weit * wbce).sum(dim=(2, 3)) / weit.sum(dim=(2, 3))

    pred = torch.sigmoid(pred)
    inter = ((pred * mask) * weit).sum(dim=(2, 3))
    union = ((pred + mask) * weit).sum(dim=(2, 3))
    wiou = 1 - (inter + 1) / (union - inter + 1)
    return (wbce + wiou).mean()

if __name__ == '__main__':

    # 数据集路径
    img_root = './VDT-2048 dataset/Train/'
    out_path_PGT1 = './output_PGT1/'
    save_path_T = './modelT'
    if not os.path.exists(save_path_T):
        os.mkdir(save_path_T)

    lr = 0.0001
    batch_size = 8
    epoch = 200
    num_params = 0
    
    scaler = GradScaler()  # Initialize GradScaler for mixed precision

    # 初始化数据集和DataLoader
    data = Data(img_root)
    loader = DataLoader(data, batch_size=batch_size, shuffle=True, num_workers=8, pin_memory=True)  # 优化数据加载

    # 初始化模型
    net = Tnet().cuda()
    net.load_state_dict(torch.load('./model/final.pth'), strict=False)
    qnet = QAnet().cuda()
    qnet.load_state_dict(torch.load('./modelQA/final.pth'))

    for param in qnet.parameters():
        param.requires_grad = False

    # 优化器
    params = net.parameters()
    optimizer = torch.optim.Adam(params, lr, betas=(0.5, 0.999))

    # 打印模型参数数量
    for p in net.parameters():
        num_params += p.numel()
    print("Tnet的参数数量: {}".format(num_params))

    iter_num = len(loader)
    net.train()
    for epochi in tqdm(range(1, epoch + 1)):
        prefetcher = DataPrefetcher(loader)
        rgb, t, d, eg, label = prefetcher.next()

        r_T_loss = 0
        epoch_ave_loss = 0
        i = 0

        while rgb is not None:
            i += 1
            with torch.no_grad():
                score1_d, score1_t = qnet(rgb, t, d)

            # 使用混合精度训练
            with torch.autocast(device_type="cuda"): # Enable autocast for mixed precision
                score_eg, score3, score2, score1, score3_t, score2_t, score1_t, score3_d, score2_d, score1_d, score4_out, score3_out, score2_out, score1_out = net(
                    rgb, t, d, score1_d, score1_t)

                losseg_out = bce_loss(score_eg, eg)

                # 各种结构损失计算
                loss4_out = structure_loss(score4_out, label)
                loss3_out = structure_loss(score3_out, label)
                loss2_out = structure_loss(score2_out, label)
                loss1_out = structure_loss(score1_out, label)

                loss3 = structure_loss(score3, label)
                loss2 = structure_loss(score2, label)
                loss1 = structure_loss(score1, label)

                loss3_t = structure_loss(score3_t, label)
                loss2_t = structure_loss(score2_t, label)
                loss1_t = structure_loss(score1_t, label)

                loss3_d = structure_loss(score3_d, label)
                loss2_d = structure_loss(score2_d, label)
                loss1_d = structure_loss(score1_d, label)

                # 总的Tnet损失
                T_loss = losseg_out + loss1 + loss2 + loss3 + loss1_t + loss2_t + loss3_t + loss1_d + loss2_d + loss3_d + loss1_out * 2 + loss2_out + loss3_out + loss4_out
                r_T_loss += T_loss.item()  # 使用item()替代data

            # 使用混合精度梯度缩放进行反向传播
            scaler.scale(T_loss).backward()

            # 使用优化器更新权重
            scaler.step(optimizer)
            scaler.update()

            optimizer.zero_grad()

            if i % 100 == 0:
                print(f'epoch: [{epochi}/{epoch}], iter: [{i}/{iter_num}]  ||  loss: {r_T_loss / 100:.4f}, lr: {lr:.6f}')
                epoch_ave_loss += (r_T_loss / 100)
                r_T_loss = 0

            rgb, t, d, eg, label = prefetcher.next()

        print(f'epoch-{epochi}_ave_loss: {epoch_ave_loss / (10.5 / batch_size):.6f}')

        # 每10个epoch保存一次模型
        if epochi % 10 == 0:
            model_path = f'{save_path_T}/epoch_T_{epochi}.pth'
            torch.save(net.state_dict(), model_path)

    # 最后保存模型
    torch.save(net.state_dict(), f'{save_path_T}/final.pth')


希望得到您的指正，我将十分的感激！！！

**Comment by zzz-0930** [comment]
或者是不是我冻结qnet的方式出错了

**Comment by Lx-Bao** [comment]
您好，我使用Ttrain.py来对Tnet进行200个epoch的训练，用时是14个小时；我也尝试了在我的机器上运行您提供的修改后的训练代码并将Batchsize设置为4，预计训练时间为7小时。您的问题应该不是出在Ttrain.py的代码上，建议您检查下是否对Tnet里面的模型结构进行过修改导致显存占用和训练时间的增长？

**Comment by zzz-0930** [comment]
再次的感谢您，我怀疑是之前的cuda版本太老没有更新，所以导致Tnet的训练时间太长，昨天把model和QAmodel重新的训练了一下，惊喜的发现目前Tnet的训练需要17个半小时（用的是您给的代码以及模型参数都没变）。model的训练时间是8个多小时，QAmodel的训练时间是5个多小时，不知道这三个的训练时间是否正常。
另外我在使用torch.cuda.amp加速时，虽然速度加快了，但是发现loss变为了nan，不知道有什么方法去避免！
我是这方面的小白，多多麻烦您了！！

不对，我去训练Tnet的时候batch_size设置为4跑个2个小时后会训练时间变为40+h,我怀疑是gpu内存超了去调用共享内存了，batch_size设置为2要22h，您要是有空的话方便发下您当时使用的环境信息吗，

**Comment by Lx-Bao** [comment]
您好，model和QAmodel的训练时间是正常的。loss变为nan主要是由于半精度训练时，计算边缘损失时产生的错误，您可以考虑更换损失函数或者在计算边缘部分逐层禁用 autocast。
我的环境是 Torch 2.1.0+cuda 12.1
针对显存超出问题，您可以不使用data_prefetcher的数据预加载方法。这样在训练过程中显存就不会增加。您可以参考如下代码[[Code]](https://pan.baidu.com/s/1w4A-sTf7IFdM6S0y6eF0zg?pwd=g6bn)。建议您可以优先尝试这个方法看看是否有效。

**Comment by zzz-0930** [comment]
感谢作者的帮助，我想我以及成功复现出来了并且输出x4e_pred_dt得到以下参数
![image](https://github.com/user-attachments/assets/0242538d-d1d1-4ea9-8e46-1cfbb08838d4)
想问下这个x4e_pred_dt就是最后的输出把，在使用matlab进行评估时，发现不能直接用，便稍微修改了一下，但感觉影响不大！
再打扰下您，您论文中的fps是咋算的，我修改了model的test代码，得出来fps是4.5左右

**Comment by Lx-Bao** [comment]
您好，模型最终的显著性预测图输出为x1e_pred_dt；  x4e_pred_dt是Tnet最高层的输出，用作模型的深度监督。 对于我们的模型及所有对比模型，我们计算FPS统一采用了如下代码：
  
```
import numpy as np
import torch
import time
from tqdm import tqdm
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

def computeTime(model1,model2, device='cuda'):
    inputs1 = torch.randn(1, 3, 384, 384)
    inputs2 = torch.randn(1, 1, 384, 384)
    if device == 'cuda':
        model1 = model1.cuda()
        model2 = model2.cuda()
        inputs1 = inputs1.cuda()
        inputs2 = inputs2.cuda()

    model1.eval()
    model2.eval()

    time_spent = []
    for idx in tqdm(range(1000)):
        start_time = time.time()
        with torch.no_grad():
            _ = model1(inputs1,inputs1,inputs1)
            _ = model2(inputs1,inputs1,inputs1,inputs2,inputs2)

        if device == 'cuda':
            torch.cuda.synchronize()  # wait for cuda to finish (cuda is asynchronous!)
        if idx > 10:
            time_spent.append(time.time() - start_time)
    print('Average speed: {:.4f} fps'.format(1 / np.mean(time_spent)))


torch.backends.cudnn.benchmark = True

from QAnet import QAnet
from Tnet import Tnet

model1 = QAnet().cuda()
model2 = Tnet().cuda()

computeTime(model1,model2)
```

**Comment by zzz-0930** [comment]
您好，刚刚使用了您提供的flops文件，生成flops，发现很不稳定，从4.5到6点多都生成过
![image](https://github.com/user-attachments/assets/bd794ae3-411c-4200-a74b-75d82c4b1dca)
另外再使用x1e_pred_dt作为输出时，得到的参数不是太理想
![image](https://github.com/user-attachments/assets/568f5df3-2784-4ca0-a1ec-e9fc69835b62)
但是x4e_pred_dt作为输出时得到的参数意外的还行，我等下再重新训练一下看看是什么结果

**Comment by Lx-Bao** [comment]
您好，测试FPS时您是否是在一张空着的显卡上测试的？和其他程序共用一张显卡测试会导致FPS降低并且数值波动较大。
若您没对Tnet进行修改，您可以直接使用test_all.py进行测试，保存的预测图结果就是x1e_pred_dt。

**Comment by zzz-0930** [comment]
我是在一张没有任何其它工作的显卡上运行的，我再去试一下！
对了我想起来了，不知道是opencv的版本还是什么原因，我无法使用您提供的cv2.imwrite去进行保存，而是自己修改了一下使用pillow进行保存了，这会有影响吗
然后我的评价指标使用的是库 https://github.com/DengPingFan/CODToolbox
应该和您提供的也没区别
